icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/MeasureUnitTest.java

```diff
@@ -3282,4 +3259,198 @@ public void TestNumericTimeSomeSpecialFormats() {
+        }
+        }
+
+        TestCase cases[] = {
+                // Correctly normalized identifiers should not change
+                new TestCase("", ""),
+                new TestCase("square-meter-per-square-meter", "square-meter-per-square-meter"),
+                new TestCase("kilogram-meter-per-square-meter-square-second",
+                        "kilogram-meter-per-square-meter-square-second"),
+                new TestCase("square-mile-and-square-foot", "square-mile-and-square-foot"),
+                new TestCase("square-foot-and-square-mile", "square-foot-and-square-mile"),
+                new TestCase("per-cubic-centimeter", "per-cubic-centimeter"),
+                new TestCase("per-kilometer", "per-kilometer"),
+
+                // Normalization of power and per
+                new TestCase(
+                        "pow2-foot-and-pow2-mile", "square-foot-and-square-mile"),
+                new TestCase(
+                        "gram-square-gram-per-dekagram", "cubic-gram-per-dekagram"),
+                new TestCase(
+                        "kilogram-per-meter-per-second", "kilogram-per-meter-second"),
+
+                // TODO(ICU-20920): Add more test cases once the proper ranking is available.
+        };
+
+
+        for (TestCase testCase :cases){
+            MeasureUnit unit = MeasureUnit.forIdentifier (testCase.id);
+
+        final String actual = unit.getIdentifier();
+            assertEquals(testCase.id, testCase.normalized, actual);
+        }
+    }
+
+    @Test(expected = InternalError.class)
+    public void TestInvalidIdentifiers() {
+        final String inputs[] = {
+                "kilo",
+                "kilokilo",
+                "onekilo",
+                "meterkilo",
+                "meter-kilo",
+                "k",
+                "meter-",
+                "meter+",
+                "-meter",
+                "+meter",
+                "-kilometer",
+                "+kilometer",
+                "-pow2-meter",
+                "+pow2-meter",
+                "p2-meter",
+                "p4-meter",
+                "+",
+                "-",
+                "-mile",
+                "-and-mile",
+                "-per-mile",
+                "one",
+                "one-one",
+                "one-per-mile",
+                "one-per-cubic-centimeter",
+                "square--per-meter",
+                "metersecond", // Must have compound part in between single units
+
+                // Negative powers not supported in mixed units yet. TODO(CLDR-13701).
+                "per-hour-and-hertz",
+                "hertz-and-per-hour",
+
+                // Compound units not supported in mixed units yet. TODO(CLDR-13700).
+                "kilonewton-meter-and-newton-meter",
+
+        };
+
+        for (String input : inputs) {
+            // TODO: test the failure for each input value.
+          MeasureUnit.forIdentifier(input);
+        }
```

Comment:
@hagbard , how could I test that each test case will throw and `InternalError` exception

Reply:
Something like this should work:

```java
try {
    MeasureUnit.forIdentifier(input);
    throw new AssertionError("Should have failed to parse");
} catch (IllegalArgumentException e) {
    // pass
}
```

Reply:
Done

Reply:
This review was written on commit a401a97. You then force-pushed to 0bc0976 throwing away commit a401a97. Once you've sent a PR for review, please be very careful with force-pushes: if I'm reviewing, I want to be able to compare changes to my last review point, I don't want it thrown away / don't want to have to keep a copy in my own repository just in case.

We do eventually squash commits in ICU: particularly when someone says "I approve. Please squash." And I also see people squashing before then (e.g. I saw hagbard doing this for a small PR - but small PRs are easy to follow). Big commits like these need care.

In my PR 1284 I did a lot of force-pushing too: my reason was because it was the best way I knew of to test small changes on MSVC, and I was very careful about which commits I squashed. (I had "fixup" commits, and I paid attention to the commits at which people last reviewed.)